src/hardware/shared_adc.rs

```diff
@@ -0,0 +1,113 @@
+/// Shared Internal ADC Support
+///
+/// # Description
+/// This module provides an abstraction to share ownership of a single ADC peripheral with multiple
+/// ADC channels attached to it.
+///
+/// The design of this module mimics that of [`shared-bus`].
+///
+/// First, the shared ADC is created with the use of a macro, which places the ADC peripheral into
+/// a mutable, static (singleton) location. Then, individual channels are created by passing in the
+/// associated ADC input pin to the [SharedAdc::create_channel()] function to generate an
+/// [AdcChannel]. The [AdcChannel]'s ownership can then be moved to any required drivers.
+///
+/// ## Synchronization
+/// If the multiple priorities utilize the ADC that results in resource pre-emption, pre-emption is
+/// protected against through the use of an atomic bool. Attempting to utilize the ADC from a
+/// higher priority level while it is in use at a lower level will result in a [AdcError::InUse].
+use embedded_hal::adc::{Channel, OneShot};
+use stm32h7xx_hal as hal;
+
+#[derive(Debug, Copy, Clone)]
+pub enum AdcError {
+    /// Indicates that the ADC channel has already been allocated.
+    Allocated,
+
+    /// Indicates that the ADC is already in use
+    InUse,
+}
+
+/// A single channel on an ADC peripheral.
+pub struct AdcChannel<'a, Adc, PIN> {
+    pin: PIN,
+    slope: f32,
+    mutex: &'a spinning::Mutex<hal::adc::Adc<Adc, hal::adc::Enabled>>,
+}
+
+impl<'a, Adc, PIN> AdcChannel<'a, Adc, PIN>
+where
+    PIN: Channel<Adc, ID = u8>,
+    hal::adc::Adc<Adc, hal::adc::Enabled>: OneShot<Adc, u32, PIN>,
+    <hal::adc::Adc<Adc, hal::adc::Enabled> as OneShot<Adc, u32, PIN>>::Error:
+        core::fmt::Debug,
+{
+    /// Read the ADC channel.
+    ///
+    /// # Returns
+    /// The normalized ADC measurement as a ratio of full-scale.
+    pub fn read(&mut self) -> Result<f32, AdcError> {
+        let mut adc = self.mutex.try_lock().ok_or(AdcError::InUse)?;
+        Ok(adc.read(&mut self.pin).unwrap() as f32 / self.slope)
```

Comment:
Why not just

```suggestion
        Ok(adc.read(&mut self.pin).unwrap() as f32 / adc.slope() as f32)
```

And forget about `self.slope`? `adc.slope() as f32` is fast.

Reply:
Ah. That's only implemented in a macro on each `Adc<$ADC, _>` in the hal. Unfortunate.

Reply:
Yeah, its hidden behind a trait implementation and is _exceptionally_ not obvious. Took me a while to debug why it wasn't working in the first place, since I had it like this originally.